### PR TITLE
Make `bytes` an IO primitive

### PR DESCRIPTION
**Pull request by e-kayrakli**
Makes `bytes` type an IO primitive by adding support for writing and reading `bytes` to channels directly. Also adds proper `writeThis` and `readThis` that just gives compiler errors.

While there:
- Removes bytes concatenation with other types, similar to the recent change in strings
- Adds `bytes.numBytes` similar to recent change in strings
- Makes adjustments in tests and `bytes` implementation to adhere to these changes.
- Some whitespace adjustments in the `bytes` implementation.

Test status:
- [x] quickstart

### CODE REVIEW

**Comment by e-kayrakli** [comment]
@mppf will probably up for review for this.

Here are some questions I have in mind:

- Do we want to be able to read/write `bytes` to text channels?

  This PR currently assumes the answer is yes. This mainly helps in writing to `stdout` without any extra step.

- Do we need `qio_channel_*_binary_string` functions? 

  `qio_channel_*_string` functions are currently called with `bytes` object only when the format indicates binary. So, it seems like they should function fine. But, OTOH, I am not very confident in this.

- I have added read/scan support just to keep things compiling. Should we just halt for now? Or should I add some tests and have complete IO support for the bytes type?

- Should these changes be tested on any particular config? Or a quickstart is enough?

**Comment by mppf** [comment]
> Here are some questions I have in mind:
> 
> - Do we want to be able to read/write `bytes` to text channels?
> 
>   This PR currently assumes the answer is yes. This mainly helps in writing to `stdout` without any extra step.

I think "yes" as well.

> 
> - Do we need `qio_channel_*_binary_string` functions? 
> 
>   `qio_channel_*_string` functions are currently called with `bytes` object only when the format indicates binary. So, it seems like they should function fine. But, OTOH, I am not very confident in this.

There are already 2 functions: qio_channel_write_string and qio_channel_print_string. The write_string one is for binary data, so can apply to `bytes`. So:
 * writing a string to a text channel calls qio_channel_print_string
 * writing a bytes to a text channel calls qio_channel_write_string
 * writing a string to a binary channel calls qio_channel_write_string
 * writing a bytes to a binary channel calls qio_channel_write_string

It's possible something in qio_channel_write_string sometimes needs to know if the string is UTF-8 encoded or not, but AFAIK it does not. Same situation for reading.

> - I have added read/scan support just to keep things compiling. Should we just halt for now? Or should I add some tests and have complete IO support for the bytes type?

I think it's acceptable to halt for this PR but I think we should expect to have reading done for the release. It should more or less amount to calling qio_channel_read_string.

> - Should these changes be tested on any particular config? Or a quickstart is enough?

We normally test each PR on the "standard" configuration and that's probably enough. Quickstart is fine during development (e.g. if you want to check testing before asking for review).

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -2728,6 +2729,13 @@ private proc _read_text_internal(_channel_internal:qio_channel_ptr_t,
     var ret = qio_channel_scan_string(false, _channel_internal, tx, len, -1);
     x = new string(tx, length=len, needToCopy=false);
     return ret;
+  } else if t == _bytes {
+    // handle _bytes
+    var len:int(64);
+    var tx: c_string;
+    var ret = qio_channel_scan_string(false, _channel_internal, tx, len, -1);
```
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/IO.chpl]: Oh, I see, the qio_channel_scan_string / qio_channel_print_string handle various forms of escaping. Perhaps it is reasonable for these to do that escaping for bytes strings too.

 * if we want writef("%ht", bytesContainingOneZeroByte) to emit an escaped `\x00` or `\u0000` then it should call qio_channel_scan_string / qio_channel_print_string and you should add an argument to indicate if the string is UTF-8 or bytes. If it's bytes, you should read adjust the method to not call read_char / write_char (these would do character-set conversion) and rather call qio_channel_read_byte / qio_channel_write_byte (or just read/write with a 1 byte buffer).
 * if we want writef("%ht", bytesContainingOneZeroByte) to emit just the null byte, then it should call qio_channel_read_string / qio_channel_write_string, and we should do our best in the formatted I/O code to make sure that the various format details (padded width, etc) apply reasonably well to the bytes.

[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/IO.chpl]: I think direction-wise, I want the first option. As, in general, it'd be helpful to printout `bytes` without creating a huge mess because of non-conforming bytes. In such cases we should instead use `\x00` or similar. I will see if I can get this right in the implementation.

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -2775,6 +2783,10 @@ private proc _write_text_internal(_channel_internal:qio_channel_ptr_t,
     // handle string
     const local_x = x.localize();
     return qio_channel_print_string(false, _channel_internal, local_x.c_str(), local_x.numBytes:ssize_t);
+  } else if t == _bytes {
+    // handle bytes
+    const local_x = x.localize();
+    return qio_channel_print_string(false, _channel_internal, local_x.c_str(), local_x.numBytes:ssize_t);
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/IO.chpl]: .. that is if we don't want to handle escapes in bytes, right. I thought we would. 
e.g don't we want `"AB\nC":bytes` to be printed as 
```
AB
C
```

File: modules/standard/IO.chpl [hunk 3]
```diff
@@ -5321,14 +5345,37 @@ proc _toNumeric(x:?t) where !_isIoPrimitiveType(t)
...
+}
 
 private inline
 proc _toString(x:?t) where t==string
 {
   return (x, true);
 }
 private inline
-proc _toString(x:?t) where (_isIoPrimitiveType(t) && t!=string)
+proc _toString(x:?t) where t==_bytes
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/IO.chpl]: I can't think of any reason why I wrote it that way. Will change.

File: modules/standard/IO.chpl [hunk 4]
```diff
@@ -5390,7 +5437,12 @@ proc _setIfPrimitive(ref lhs:?t, rhs:?t2, argi:int):syserr where t!=bool&&_isIoP
         lhs = rhs:int:t;
       }
     } else {
-      lhs = rhs:t;
+      if isBytesType(t2) && isStringType(t) {
+        lhs = rhs.decode(DecodePolicy.Ignore);
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/IO.chpl]: My thinking was what happens here pertains to UTF-8 validation, and since it is not implemented yet (as far as I understand), we could simply ignore non-UTF8 characters for now. But I will change it to Strict and see if any issue comes up

File: modules/standard/IO.chpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/IO.chpl]: Right, I was getting confused. we have to make a choice but I think the other comments covered it.

File: modules/standard/IO.chpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/IO.chpl]: I thought that "Ignore" meant "drop these characters" rather than "put them in string unmodified"

[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/IO.chpl]: Oops, I got confused. It makes me think, though. May be we need that policy, too. It'd simply amount to copying the buffer entirely without any check. But maybe that shouldn't happen in `decode`.

For this particular case, I agree that `Strict` is more appropriate than `Ignore`

File: modules/standard/IO.chpl [hunk 5]
```diff
@@ -2728,6 +2729,9 @@ private proc _read_text_internal(_channel_internal:qio_channel_ptr_t,
     var ret = qio_channel_scan_string(false, _channel_internal, tx, len, -1);
     x = new string(tx, length=len, needToCopy=false);
     return ret;
+  } else if t == _bytes {
+    // handle _bytes
+    halt("Cannot scan into bytes objects yet");
```
Comment (mppf) [inline modules/standard/IO.chpl]: This is presumably temporary, but why `halt` here? You could set `err = EINVAL`.

File: modules/standard/IO.chpl [hunk 6]
```diff
@@ -2860,6 +2868,9 @@ private inline proc _read_binary_internal(_channel_internal:qio_channel_ptr_t, p
                                       _channel_internal, tx, len, -1);
     x = new string(tx, length=len, needToCopy=false);
     return ret;
+  } else if t == _bytes {
+    // handle _bytes
+    halt("Cannot read into bytes objects yet");
```
Comment (mppf) [inline modules/standard/IO.chpl]: This case is easy in the binary channel, isn't it? AFAIK qio_channel_read_string doesn't assume anything about UTF-8. Nonetheless it's OK to do it in another PR.

File: modules/standard/IO.chpl [hunk 7]
```diff
@@ -2975,7 +2989,8 @@ private inline proc _read_one_internal(_channel_internal:qio_channel_ptr_t,
 private inline proc _write_one_internal(_channel_internal:qio_channel_ptr_t,
                                         param kind:iokind,
                                         x:?t,
-                                        loc:locale?):syserr where _isIoPrimitiveTypeOrNewline(t) {
+                                        loc:locale?,
+                                        byte_esc=false):syserr where _isIoPrimitiveTypeOrNewline(t) {
```
Comment (mppf) [inline modules/standard/IO.chpl]: I might be missing / misunderstanding some case, but it seems to me that `byte_esc` should not be an argument to this function and instead it should consider `byte_esc=true` whenever `t==_bytes`.

File: runtime/src/qio/qio_formatted.c [hunk 8]
```diff
@@ -1044,6 +1044,37 @@ char _qio_tohex(unsigned char i)
...
+      b == '\'' || b == '"' || b == '\n' ) {
+    tmp[0] = '\\';
+    if( b == '\n' ) {
+      tmpchr = 'n';
+    } else {
+      tmpchr = b;
+    }
+    tmp[1] = tmpchr;
+    cwidth = 2;
+  } else if( !iswascii(b) || !isprint(b) ) {
```
Comment (mppf) [inline runtime/src/qio/qio_formatted.c]: I've never heard of iswascii and my system doesn't have a `man` for it. Can you instead simply check `b >= 128`?

File: test/types/bytes/bytesFormattedIO.chpl [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+writeln();
+writef("<%|4S>\n", b);
+writeln();
+writef("<%|8S>\n", b);
+writeln();
+writef("<%|*S>\n", 1, b);
+writeln();
+writef("<%|vS>\n", b);
+writeln();
+writef("<%ht>\n", b);
```
Comment (mppf) [inline test/types/bytes/bytesFormattedIO.chpl]: Can you add some %s arguments with things like field widths? Maybe they fail today... but it would be good to have testing of them.

Comment (mppf) [reply test/types/bytes/bytesFormattedIO.chpl]: Also it'd be nice to have the text tests in a separate test, so that the output .good can more easily be examined for that case.

**Comment by e-kayrakli** [comment]
OK, I am able to remove the `byte_esc` argument. The resulting design is as follows:

- When `writef` is called with binary string conversion type, both `bytes` and `string` types are converted to `string`. Even if a `bytes` is used and it had some non-UTF8 characters, we are not going to do anything about it here anyways.
- `_write_one_internal` is unchanged.
- `_write_text_internal` calls `qio_channel_print_bytes` if called with a `bytes`.
- The ugly `byte_esc` argument is only used internally in the runtime as a switch to `qio_channel_print_string_or_bytes` which is now a static function called by `qio_channel_print_bytes` and `qio_channel_print_string`

I think other smaller issues have also been addressed by the last commits. Full testing passed with standard config, now running gasnet

**Comment by mppf** [comment]
Thanks for the updates, I think it can merge now.

> When writef is called with binary string conversion type, both bytes and string types are converted to string. Even if a bytes is used and it had some non-UTF8 characters, we are not going to do anything about it here anyways.

The issue here is that we'll run into validation errors for bytes containing non-UTF8. It seems wiser to me to convert them both to `bytes` or possibly even `c_str` + a length.

Anyway this PR can merge with this issue, just be sure to call it out as an issue so we can attend to it once we start validating UTF-8.
